### PR TITLE
Revamp the engine style guide, remove `always_specify_types`.

### PR DESCRIPTION
**Pull request by matanlurey**
Based on the (internal) discussion around converging on using the official Dart style guide, with the exception of the code that gets published under `dart:ui`, as that is user-facing, and we'd like to evolve the code style in conjunction with the framework.

I also took the opportunity to specify more about our style guide use in general, mostly to make it easier to understand our conventions, and also call out known problem areas (notably, our over-use of `shared_ptr` and `auto` in some cases). I am happy to split those up, but it was easier to make the markdown changes at once.

I also took @cbracken and folks advice and clarified directly that explicit types in Dart are _not_ bad (with examples).

### CODE REVIEW

File: analysis_options.yaml [hunk 1]
```diff
@@ -29,7 +29,7 @@ linter:
     - always_declare_return_types
     - always_put_control_body_on_new_line
     # - always_put_required_named_parameters_first # we prefer having parameters in the same order as fields https://github.com/flutter/flutter/issues/10219
-    - always_specify_types
+    # - always_specify_types # no longer used; see https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#dart
```
Comment (goderbauer) [inline analysis_options.yaml:32]: Please mark this as described at the top of the file.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -14,29 +14,177 @@ contributing guide.
...
+
+- Use C++ casts (C-style casts are forbidden).
+- Use `nullptr` rather than `NULL`.
+- Avoid `#define`; for internal constants use `static constexpr` instead.
+
+### Dart
+
+The Flutter engine _intends_ to follow the [Dart style guide][dart_style] but
+currently follows the [Flutter style guide][flutter_style], with the following
+exceptions:
```
Comment (goderbauer) [inline CONTRIBUTING.md:79]: I'd like to learn more about why it was decided to make that shift and how it'll benefit us to have diverging style guides in the engine and framework. Where can I learn more?

**Comment by Hixie** [comment]
Before making this kind of dramatic change we should have a proper discussion with a design doc and Dash Forum discussion and so on.

**Review by Hixie — CHANGES_REQUESTED** [review]
Before making this kind of dramatic change we should have a proper discussion with a design doc and Dash Forum discussion and so on. There's lots of tradeoffs to consider.

File: CONTRIBUTING.md (hunk 2 above)
Comment (johnmccutchan) [reply CONTRIBUTING.md:79]: The root problem is that we _already_ have diverging style guides: Dart vs. Flutter.

As a project, Dash needs to settle on a single style guide (and automatic formatter) for every language we use.

**Comment by johnmccutchan** [comment]
The folks who work day-to-day in the engine repository have decided that they want to remove `always_specify_types` for two reasons:

1) They feel that the ergonomics will be improved for Dart code that they write in the engine repository.
2) We have a larger plan to standardize the Dash project on one style per language and disabling this lint is inline with adopting Dart's style guide.

I (and I imagine the rest of the engine team) am interested in hearing any feedback from people who regularly contribute to the engine repository. 

Feedback from people is still appreciated, but, it will be scaled accordingly.

**Review by johnmccutchan — APPROVED** [review]

**Comment by matanlurey** [comment]
Fwiw, I'd be happy to keep the style guide in place on `dart:ui` (`lib/ui`) for the time being.

The _vast_ majority of Dart code in the engine is custom tooling that is not exposed to end-users at all.

**Comment by Hixie** [comment]
I think it's very important that we take into account the needs of our team, certainly, including especially folks directly working on the relevant code. It's not the only input, however, and as the project's tech lead I am requesting that you not unilaterally change long-established project-wide policies without presenting the needs that are leading to those desired changes, objectively considering the needs of our users, and generally working in the open way that Flutter's core values are all about.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -14,29 +14,177 @@ contributing guide.
...
+
+#### Use of type inference is allowed
+
+The [Dart style guide][dart_inference] only requires explicit types when type
+inference is not possible, but the Flutter style guide always requires explicit
+types. The engine is moving towards the Dart style guide, but this is a gradual
+process. In the meantime, follow these guidelines:
+
+- **Always** annotate when inference is not possible.
```
Comment (cbracken) [inline CONTRIBUTING.md:88]: So long as implicit dynamic is disallowed, it seems unnecessary to state this, though I'm fine making this extra explicit even if so.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -14,29 +14,177 @@ contributing guide.
...
+- **Consider** annotating when inference is possible but the type is not
+  obvious.
+- **Prefer** annotating when a code review requests it, regardless.
```
Comment (cbracken) [inline CONTRIBUTING.md:91]: My personal preference would be to prefer annotating when inference is possible but the type is unobvious. I think it's unnecessary to mention to prefer annotating when a reviewer requests it -- that's a decision between the author and reviewer, based on the guidance in the style guide.

In my experience, much of the time reading code is not in an IDE -- it's in code reviews, browsing the repo, grepping the source, etc.

**Comment by cbracken** [comment]
To be clear, my personal position is that I _strongly prefer_ that the type of a
given expression be obvious. I find it difficult to read and review code where
the types are unobvious; types aren't just for the compiler -- they're very useful to readers in helping them reason about the code.

Forcing explicitly-typed declarations is one approach to making the types obvious, but I agree it often does more harm to readability than good. As an example, in [`platform_isolate.dart`](https://github.com/flutter/engine/blob/14d27622b3144d1e14218cbb789846b00360416a/lib/ui/platform_isolate.dart#L48) we have this gem:
```dart
final Map<int, Completer<Object?>> _pending = <int, Completer<Object?>>{};
```
which seems far less readable than:
```dart
final _pending = <int, Completer<Object?>>{};
```

Taking it a step further, the final line seems obvious enough to not need a type at all:
```dart
final logBuffer = StringBuffer();
// ... add messages
final log = logBuffer.toString();
```

Overall, I'm in favour of having the style guide include explicit guidance that the types be obvious to the reader, but leaving the actual implementation of that obviousness to a discussion between the author and the reviewer rather than enforcing brute-force via tooling.

**Review by cbracken — APPROVED** [review]
Post-update, the issues I had are all addressed.

**Comment by zanderso** [comment]
Reason for revert: This sort of change *is* in scope for a discussion at the Dash forum.

**Comment by Hixie** [comment]
Landing a change while there is an outstanding review comment is a [violation of our policies](https://github.com/flutter/flutter/wiki/Tree-hygiene#who) and _extremely_ rude, which also makes it a violation of our code of conduct.

**Pull request merged**